lib/collection/src/shard/shard_holder.rs

```diff
@@ -0,0 +1,236 @@
+
+    pub fn start_shard_transfer(
+        &mut self,
+        shard_id: ShardId,
+        to_peer: PeerId,
+        this_peer: PeerId,
+    ) -> CollectionResult<ShardTransfer> {
+        let shard = self.shards.get(&shard_id).ok_or_else(|| {
+            CollectionError::service_error("Shard {shard_id} is absent".to_owned())
+        })?;
+        let from_peer = shard.peer_id(this_peer);
+        let shard_transfer = ShardTransfer {
+            from: from_peer,
+            to: to_peer,
+        };
+        self.shard_transfers
+            .insert(shard_id, shard_transfer.clone());
+        Ok(shard_transfer)
+    }
+
+    pub fn finish_transfer(&mut self, shard_id: ShardId) -> CollectionResult<()> {
+        let transfer = self.shard_transfers.remove(&shard_id).ok_or_else(|| {
+            CollectionError::service_error(format!(
+                "Shard transfer data for {shard_id} is absent at the end of the transfer."
+            ))
+        })?;
+        if let Shard::Remote(shard) = self.shards.get_mut(&shard_id).ok_or_else(|| {
+            CollectionError::service_error("Shard {shard_id} is absent".to_owned())
+        })? {
+            shard.peer_id = transfer.to;
+        };
+        Ok(())
+    }
+
+    pub fn local_shard_by_id(&self, id: ShardId) -> CollectionResult<&Shard> {
+        match self.shards.get(&id) {
+            None => Err(CollectionError::bad_shard_selection(format!(
+                "Shard {} does not exist",
+                id
+            ))),
+            Some(Shard::Remote(_)) => Err(CollectionError::bad_shard_selection(format!(
+                "Shard {} is not local on peer",
+                id
+            ))),
+            Some(shard @ Shard::Local(_)) => Ok(shard),
+            Some(shard @ Shard::Proxy(_)) => Ok(shard),
+        }
+    }
+
+    pub fn target_shards(&self, shard_selection: Option<ShardId>) -> CollectionResult<Vec<&Shard>> {
+        match shard_selection {
+            None => Ok(self.all_shards().collect()),
+            Some(shard_selection) => {
+                let local_shard = self.local_shard_by_id(shard_selection)?;
+                Ok(vec![local_shard])
+            }
+        }
+    }
+
+    pub async fn before_drop(&mut self) {
+        let futures: FuturesUnordered<_> = self
+            .shards
+            .iter_mut()
+            .map(|(_, shard)| shard.before_drop())
+            .collect();
+        futures.collect::<Vec<()>>().await;
+    }
+
+    pub fn len(&self) -> usize {
+        self.shards.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.shards.is_empty()
+    }
+}
+
+impl LockedShardHolder {
+    pub fn new(shard_holder: ShardHolder) -> Self {
+        Self(RwLock::new(shard_holder))
+    }
+
+    pub async fn add_shard(&self, shard_id: ShardId, shard: Shard) {
+        let mut holder = self.0.write().await;
+        holder.add_shard(shard_id, shard).await;
+    }
+
+    pub async fn remove_shard(&self, shard_id: ShardId) -> Option<Shard> {
```

Comment:
Most of the locked shard holder impl code can be removed by implementing `Deref<Target=RwLock<...>>` for it. It is the exact case for which `Deref` is used. You can check we have similar impls with Deref in our codebase.

Reply:
https://doc.rust-lang.org/std/ops/trait.Deref.html

Reply:
Actually, instead of using deref, I think, I better remove this methods on a locked container. We are not forcing to use `LockedShardHolder` interface to interact with shards, so it might be a potential place to accidentally have a double read lock.